Update TXT.php

MasahikoSada opened the pull request:
To support files encoded by Shift-JIS(Japanese) , GB2312 and GBK(Chinese) .

MasahikoSada committed:
Update TXT.php

To support encoded by Shift-JIS(Japanese) , GB2312 and GBK(Chinese) files

erhuz reviewed (changes requested):
Would be nice to indent row 70 and remove an indentation from row 69.

kesselb commented:
Thanks :+1: 

I would say the issue is not the encoding but the font used for txt preview NotoSans-Regular.ttf

On the other hand imagettftext expects a utf8 encoded string as text anyway so a explicit is probably not wrong. Yet https://www.php.net/manual/en/function.mb-convert-encoding should be easier to use here as the from encoding is detected automatically.

kesselb commented on lib/private/Preview/TXT.php:
```diff
@@ -64,6 +64,12 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
 			return null;
 		}
+		
+		$encoding = mb_detect_encoding($content . 'a', 'SJIS-win, GB2312, GBK, UTF-8, WINDOWS-1252, ISO-8859-15, ISO-8859-1, ASCII', true);
+            if ($encoding === '') {
+				$encoding = 'ISO-8859-1';
+				}
```
```suggestion
```

kesselb commented on lib/private/Preview/TXT.php:
```diff
@@ -64,6 +64,12 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+            if ($encoding === '') {
+				$encoding = 'ISO-8859-1';
+				}
+		$content = iconv($encoding, 'UTF-8', $content);
```
```suggestion
		$content = mb_convert_encoding($content, 'UTF-8');
```

MasahikoSada committed:
Update lib/private/Preview/TXT.php

Thanks, I appreciate your suggestion.

Co-authored-by: kesselb <mail@danielkesselberg.de>

MasahikoSada committed:
Update lib/private/Preview/TXT.php

Thanks, I appereate your suggestion :)

Co-authored-by: kesselb <mail@danielkesselberg.de>

MasahikoSada committed:
Update TXT.php

In my environment, mb_convert_encoding() doesn't work properly without 'auto' in from_encoding.

kesselb commented:
Thanks @MasahikoSada, can you provide a simple text document or  a string with some characters to test the change?

MasahikoSada commented:
> Thanks @MasahikoSada, can you provide a simple text document or a string with some characters to test the change?

@kesselb 
Thanks for your reply.

This is a sample file endoced by SHIFT-JIS(ANSI).
https://1drv.ms/t/s!AnT4d94afUwMhYJsm8nD3x9bDGas8w?e=LDAdM7

kesselb commented:
Hi,

uploaded the test file to my dev instance to generate the preview images. Is one of the previews the expected outcome?

### 1. no patch

![txt_master](https://user-images.githubusercontent.com/3902676/131253060-457c8b44-afee-4ac4-9247-00e5434c6664.png)

### 2. version with iconv

![txt_iconv](https://user-images.githubusercontent.com/3902676/131253083-caa0d74e-5ddf-4ff0-ba5f-5e929fab524b.png)

### 3. version with mb_convert_encoding

![txt_mb](https://user-images.githubusercontent.com/3902676/131253111-31aba55d-28a3-4fba-a90e-827ef2b0f85d.png)

MasahikoSada commented:
> Hi,
> 
> uploaded the test file to my dev instance to generate the preview images. Is one of the previews the expected outcome?
> 
> ### 1. no patch
> ![txt_master](https://user-images.githubusercontent.com/3902676/131253060-457c8b44-afee-4ac4-9247-00e5434c6664.png)
> 
> ### 2. version with iconv
> ![txt_iconv](https://user-images.githubusercontent.com/3902676/131253083-caa0d74e-5ddf-4ff0-ba5f-5e929fab524b.png)
> 
> ### 3. version with mb_convert_encoding
> ![txt_mb](https://user-images.githubusercontent.com/3902676/131253111-31aba55d-28a3-4fba-a90e-827ef2b0f85d.png)

Hi, 
No one is the expected output, but I think that the version 2's encode is proper one.
The only issue of version 2 is just the font.

Can you show me the version  with mb_convert_encoding and 'auto' option.
I think it will be as same as version 2.

kesselb commented:
Hey,

Input `���肪�Ƃ��B
����͓��{�ꕶ���̃e�X�g�t�@�C���ł��B`

mb_convert_encoding($content, 'UTF-8', 'auto'); =>  `false`

Blank preview image

mb_convert_encoding($content, 'UTF-8', 'SJIS-win, GB2312, GBK, UTF-8, WINDOWS-1252, ISO-8859-15, ISO-8859-1, ASCII') => `ありがとう。
これは日本語文字のテストファイルです。`

![txt_mb_list_of_encodings](https://user-images.githubusercontent.com/3902676/131261194-c706307d-433d-406a-90fa-1c635b8d25c4.png)

MasahikoSada commented:
> Hey,
> 
> $content = `���肪�Ƃ��B ����͓��{�ꕶ���̃e�X�g�t�@�C���ł��B`
> 
> mb_convert_encoding($content, 'UTF-8', 'auto'); => `false`
> 
> Blank preview image
> 
> mb_convert_encoding($content, 'UTF-8', 'SJIS-win, GB2312, GBK, UTF-8, WINDOWS-1252, ISO-8859-15, ISO-8859-1, ASCII') => `ありがとう。 これは日本語文字のテストファイルです。`
> 

Thanks for your comment.
When 'auto' is set, I think it depends on 'mbstring.language'.
When I set mbstring.language=japanese, the output is expected one.
